container/snapshot.go

```diff
@@ -0,0 +1,152 @@
+package container
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/go-connections/nat"
+)
+
+// Snapshot is a read only view for Containers
+type Snapshot struct {
+	ID           string `json:"Id"`
+	Name         string
+	Pid          int
+	Managed      bool
+	Image        string
+	ImageID      string
+	Command      string
+	Ports        []types.Port
+	ExposedPorts nat.PortSet
+	PublishPorts nat.PortSet
+	Labels       map[string]string
+	State        string
+	Status       string
+	Health       string
+	HostConfig   struct {
+		NetworkMode string
+		Isolation   string
+	}
+	NetworkSettings types.SummaryNetworkSettings
+	Mounts          []types.MountPoint
+	Created         time.Time
+	StartedAt       time.Time
+	Running         bool
+	Paused          bool
+	ExitCode        int
+}
+
+// Snapshot provides a read only view of a Container. Callers must hold a Lock on the container object.
+func (container *Container) Snapshot() *Snapshot {
```

Comment:
It is similar, but not necessarily the same. It's more of a mapping to what queries need to see, rather than only a (deep) copy. The rest of the code already has precedence for similar mapping to different views/strucs being done at different layers.

It is fragile in the same sense that public API handlers can break when new fields are added and not properly mapped to strucs being serialized for responses.